lib/puppet/application.rb

```diff
@@ -281,6 +281,15 @@ def run_mode( mode_name = nil)
       @run_mode = Puppet::Util::RunMode[ mode_name || Puppet.settings.preferred_run_mode ]
     end
 
+    def environment_mode(mode_name)
+      raise "Invalid environment mode '#{mode_name}'" unless [:local, :remote, :not_required].include?(mode_name)
```

Comment:
@Magisus what's the best way to raise errors? It looks like there are a whole bunch of utilities around surfacing errors. Is there any standardization about which to use when?

Reply:
Hah, not that I've seen. That's part of what made wrapping those strings for translation so hard. On that note, this string should be wrapped in `_()` and have the interpolation replaced with format string syntax, whichever error raising method you end up using. See https://github.com/puppetlabs/puppet/blob/master/CONTRIBUTING.md#writing-translatable-code for details.

Reply:
It's an error only seen by people writing Applications that misuse `environment_mode`. I'd be comfortable with either `Puppet::Error`, similar to other things in Application.